File: DataBinding/PageAndRDL/ODataDataSource/VB.NET/ObjectDataSourceClient/MainForm.vb
Imports System.IO
Imports GrapeCity.ActiveReports
Imports ODataDataSource.ODataDataSource.Default

''' <summary>
''' ObjectDataSource client, which get data via OData
''' </summary>
Public Class MainForm
	Private ReadOnly _dataList As IList(Of Year)
	Private Const serviceUri As String = "http://localhost:55856/"

	Public Sub New()
		InitializeComponent()
		Dim container = New Container(New Uri(serviceUri))
		_dataList = GetYears(container)
	End Sub

	Private Function GetYears(container As Container) As IList(Of Year)
		Dim years = New List(Of Year)()
		For Each movie In container.Movies
			Dim year = movie.YearReleased
			If years.Count = 0 OrElse years(years.Count - 1).YearReleased <> year Then
				years.Add(New Year(year))
			End If
			Dim newMovie = New Movie(movie.Id, movie.Title, movie.MPAA)
			years(years.Count - 1).Movies.Add(newMovie)
		Next
		Return years
	End Function

	Private Sub OnLocateDataSource(sender As Object, args As LocateDataSourceEventArgs)
		If args.DataSet.Name = "Years" Then
			' ObjectsReport :bind collection to report
			args.Data = _dataList
		ElseIf args.DataSet.Name = "Movies" Then
			' SubObjectsReport :bind subcollection to subreport
			Dim yearReleased = 1970
			For Each param In args.Parameters
				If param.Name = "YearReleased" Then
					yearReleased = Integer.Parse(param.Value.ToString())
					Exit For
				End If
			Next
			For i As Integer = 0 To _dataList.Count - 1
				If _dataList(i).YearReleased = yearReleased Then
					args.Data = _dataList(i).Movies
				End If
			Next
		End If
	End Sub

	Private Sub MainForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
		Dim rptPath = New FileInfo("..\..\..\..\Reports\ObjectsReport.rdlx")
		Dim pageReport = New PageReport(rptPath)
		AddHandler pageReport.Document.LocateDataSource, AddressOf OnLocateDataSource
		reportPreview.LoadDocument(pageReport.Document)
	End Sub
End Class

File: DataBinding/PageAndRDL/ODataDataSource/C#/ObjectDataSourceClient/MainForm.cs
using GrapeCity.ActiveReports;
using ObjectDataSource.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Windows.Forms;

namespace ODataDataSource
{
	/// <summary>
	/// Client is based on article https://docs.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/odata-v4/create-an-odata-v4-client-app
	/// </summary>
	public partial class MainForm : Form
	{
		private readonly IList<Year> _dataList;
		private const string serviceUri = "http://localhost:55749/";
		public MainForm()
		{
			InitializeComponent();
			var container = new ObjectDataSource.Default.Container(new Uri(serviceUri));
			_dataList = GetYears(container);
		}

		private IList<Year> GetYears(ObjectDataSource.Default.Container container)
		{
			var years = new List<Year>();
			foreach (var movie in container.Movies)
			{
				int year = movie.YearReleased;
				if (years.Count == 0 || years[years.Count - 1].YearReleased != year)
					years.Add(new Year(year));
				var newMovie = new Movie(movie.Id, movie.Title, movie.MPAA);
				years[years.Count - 1].Movies.Add(newMovie);
			}
			return years;
		}

		private void OnLocateDataSource(object sender, LocateDataSourceEventArgs args)
		{
			if (args.DataSet.Name == "Years") // ObjectsReport :bind collection to report
			{
				args.Data = _dataList;
			}
			else if (args.DataSet.Name == "Movies") // SubObjectsReport :bind subcollection to subreport
			{
				var yearReleased = 1970;
				foreach (var param in args.Parameters)
				{
					if (param.Name == "YearReleased")
					{
						yearReleased = int.Parse(param.Value.ToString());
						break;
					}
				}

				for (int i = 0; i < _dataList.Count; i++)
					if (_dataList[i].YearReleased == yearReleased)
					{
						args.Data = _dataList[i].Movies;
						break;
					}
			}
		}

		private void MainForm_Load(object sender, EventArgs e)
		{
			var rptPath = new FileInfo(@"..\..\..\..\Reports\ObjectsReport.rdlx");
			var pageReport = new PageReport(rptPath);
			pageReport.Document.LocateDataSource += new LocateDataSourceEventHandler (OnLocateDataSource);
			reportPreview.LoadDocument(pageReport.Document);
		}
	}
}
